File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/gemstone/ContactStore.kt
package com.gemwallet.android.data.repositories.gemstone

import com.gemwallet.android.data.service.store.database.ContactsDao
import com.gemwallet.android.data.service.store.database.entities.toModel
import com.gemwallet.android.data.service.store.database.entities.toRecord
import com.gemwallet.android.serializer.decodeJson
import com.gemwallet.android.serializer.toJson
import com.gemwallet.android.cases.contacts.ContactRecipient
import com.wallet.core.primitives.Chain
import com.wallet.core.primitives.Contact
import com.wallet.core.primitives.ContactData
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import com.wallet.core.primitives.ContactAddress
import uniffi.gemstone.GemContactStore

class GemstoneContactStore(
    private val contactsDao: ContactsDao,
) : GemContactStore {

    override suspend fun getAddresses(contactId: String): List<String> =
        contactsDao.getAddresses(contactId).map { it.toModel().toJson() }

    override suspend fun saveContact(contact: String, addresses: List<String>) {
        contactsDao.addContact(
            contact.decodeJson<Contact>().toRecord(),
            addresses.map { it.decodeJson<ContactAddress>().toRecord() },
        )
    }

    override suspend fun updateContact(contact: String, addresses: List<String>, deleteAddressIds: List<String>) {
        contactsDao.updateContact(contact.decodeJson<Contact>().toRecord(), deleteAddressIds, addresses.map { it.decodeJson<ContactAddress>().toRecord() })
    }

    override suspend fun deleteContact(contactId: String) = contactsDao.deleteContact(contactId)

    fun observeContacts(): Flow<List<ContactData>> = contactsDao.getContacts().map { contacts -> contacts.map { it.toModel() } }

    fun observeContactRecipients(chain: Chain): Flow<List<ContactRecipient>> =
        contactsDao.getContactRecipients(chain).map { rows -> rows.map { it.toModel() } }

    suspend fun getContact(id: String): ContactData? = contactsDao.getContact(id)?.toModel()
}

File: ios/Packages/GemstoneServices/Sources/Stores/ContactStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import struct Gemstone.Contact
import struct Gemstone.ContactAddress
import protocol Gemstone.GemContactStore
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneContactStore: GemContactStore, @unchecked Sendable {
    private let store: ContactStore

    public init(store: ContactStore) {
        self.store = store
    }

    public func getAddresses(contactId: String) async throws -> [Gemstone.ContactAddress] {
        try store.getAddresses(contactId: contactId).map { $0.map() }
    }

    public func saveContact(contact: Gemstone.Contact, addresses: [Gemstone.ContactAddress]) async throws {
        try store.addContact(contact.map(), addresses: addresses.map { $0.map() })
    }

    public func updateContact(contact: Gemstone.Contact, addresses: [Gemstone.ContactAddress], deleteAddressIds: [String]) async throws {
        try store.updateContact(
            contact.map(),
            deleteAddressIds: deleteAddressIds,
            addresses: addresses.map { $0.map() },
        )
    }

    public func deleteContact(contactId: String) async throws {
        try store.deleteContact(id: contactId)
    }
}
